### PR TITLE
Add mkl implementation for exponential on CPU

### PR DESCRIPTION
**Pull request by CaoE**
### Description
Add mkl implementation for exponential on CPU to improve the performance of exponential.

### Testing
data type: float32 
single socket (28cores):
```
before: torch.Size([10, 128, 10, 124])  0.065 s
        torch.Size([10, 128, 20, 124])  0.130 s
            
after:  torch.Size([10, 128, 10, 124])  5.9e-05 s
        torch.Size([10, 128, 20, 124])  0.000113 s
```
single core:
```
before: torch.Size([10, 128, 10, 124])  0.065 s
        torch.Size([10, 128, 20, 124])  0.130 s 

after:  torch.Size([10, 128, 10, 124])  0.00117 s
        torch.Size([10, 128, 20, 124])  0.002347 s
```

cc @VitalyFedyunin @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by jbschlosser** [comment]
Hey @mruberry - not sure who the best person to review this internally was - please feel free to reassign as appropriate :)

**Comment by mruberry** [comment]
cc @VitalyFedyunin

**Comment by VitalyFedyunin** [comment]
@frank-wei PTAL

**Comment by yanbing-j** [comment]
Hi @frank-wei , could you please help review this PR?

**Comment by yanbing-j** [comment]
Hi @frank-wei , could you please help review this PR? Thank you!

**Comment by yanbing-j** [comment]
Hi @frank-wei  , could you please help review this PR? Thank you!

File: aten/src/ATen/native/cpu/DistributionKernels.cpp [hunk 1]
```diff
@@ -103,11 +103,88 @@ void bernoulli_scalar_kernel(const TensorBase &self, double p, c10::optional<Gen
...
+          // vectorized copy if using buffer and contiguous
+          if (!is_df && contig) {
+            scalar_t *self_seg = self_ptr + begin;
+            tmp_scalar_t *tmp_seg = sample_ptr + begin;
+            at::vec::convert<tmp_scalar_t, scalar_t>(tmp_seg, self_seg, len);
+          }
+        }
+      };
+
+      parallel_for(0, n, /* grain_size= */ 800, sample);
```
Comment (frank-wei) [inline aten/src/ATen/native/cpu/DistributionKernels.cpp:174]: Generally looks good to me. Does the grain_size is fixed number for any length?

Comment (CaoE) [reply aten/src/ATen/native/cpu/DistributionKernels.cpp:174]: Yes, Generally, 800 is a heuristic grain_size with relatively good performance.

**Review by frank-wei — APPROVED** [review]

**Review by frank-wei — COMMENTED** [review]
@pytorchbot merge -g

**Comment by izaitsevfb** [comment]
Heads up, @CaoE, unfortunately I have to revert this PR, as it caused breakage of Meta tests. I'll provide more details soon.

File: aten/src/ATen/native/cpu/DistributionKernels.cpp [hunk 2]
```diff
@@ -103,11 +103,88 @@ void bernoulli_scalar_kernel(const TensorBase &self, double p, c10::optional<Gen
...
 
+#if !AT_MKL_ENABLED()
+void exponential_kernel(TensorIteratorBase& iter, double lambda, c10::optional<Generator> gen) {
+  exponential_kernel_default(iter, lambda, gen);
+}
+#else
+void exponential_kernel(TensorIteratorBase &iter, double lambda, c10::optional<Generator> gen) {
+  Tensor self = iter.tensor(0);
+  if (lambda > 0 && !std::isinf(lambda) && !std::isnan(lambda) && cpuinfo_initialize() &&
+      cpuinfo_vendor_intel == cpuinfo_get_processor(0)->core->vendor) {
```
Comment (malfet) [inline aten/src/ATen/native/cpu/DistributionKernels.cpp:119]: Can somebody please explain, why this check is necessary?

**Review by malfet — CHANGES_REQUESTED** [review]
Why does it add Intel specific codepath? Wouldn't it work on all x86_64 CPUs?

**Comment by malfet** [comment]
Also, is there any documentation/results on the spatial autocorrelation / statistical properties of new RNG algorithm vs the old one?

File: aten/src/ATen/native/cpu/DistributionKernels.cpp (hunk 2 above)
Comment (jgong5) [reply aten/src/ATen/native/cpu/DistributionKernels.cpp:119]: @malfet Thanks for the ask. The same check was also applied to the `bernoulli_scalar_kernel` function (maybe added by Meta engineers in the past): https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/DistributionKernels.cpp#L50

I think the check was added because the code path leverages MKL to optimize the random distribution generation and MKL is optimized for Intel CPUs:
https://www.intel.com/content/www/us/en/developer/articles/system-requirements/oneapi-math-kernel-library-system-requirements.html

Comment (CaoE) [reply aten/src/ATen/native/cpu/DistributionKernels.cpp:119]: @jgong5 Thank you for your detailed explanation. Recently, I was occupied by other tasks and didn't reply in time.

**Comment by CaoE** [comment]
> Also, is there any documentation/results on the spatial autocorrelation / statistical properties of new RNG algorithm vs the old one?

Hi @malfet  sorry for the late reply ! I was occupied by other tasks. I will post the statistical properties of the new RNG algorithm vs the old one. Also, I didn't find relevant statistical tests on PyTorch for these distribution operators. Do we have plan to add statistical tests for these ops.

**Comment by malfet** [comment]
> Heads up, @CaoE, unfortunately I have to revert this PR, as it caused breakage of internal Meta tests. I'll provide more details soon.

To followup on the promise: after this change got landed, TestMDETRLosses::test_box_losses from https://github.com/facebookresearch/multimodal started to fail, see https://github.com/facebookresearch/multimodal/blob/afe52dec1987e5a3764539606a55f430994ebbc8/test/modules/losses/test_mdetr_losses.py#L114

Though by looking at the code it feels like it could have been a false positive (I.e. quick look at the code did not shown me where seed is fixed or where tolerance is set correctly.

**Comment by CaoE** [comment]
Hi @malfet  Tests for mean and var for exponential are added. 

> TestMDETRLosses::test_box_losses from https://github.com/facebookresearch/multimodal started to fail, see https://github.com/facebookresearch/multimodal/blob/afe52dec1987e5a3764539606a55f430994ebbc8/test/modules/losses/test_mdetr_losses.py#L114

https://github.com/facebookresearch/multimodal/blob/afe52dec1987e5a3764539606a55f430994ebbc8/test/modules/losses/test_mdetr_losses.py#L64.  `indices` for  `box_losses` is generated by multinomial which will call exponential, even the new mkl implementation fits the exponential distribution but it is not the same as the old one. Therefore,  fixed `expected_l1_loss = torch.tensor(0.7721)` and `expected_giou_loss = torch.tensor(1.1768)` is not applicable for new exponential.

Can we port the `test_box_losses` to pytorch and add more flexible `expected_l1_loss` and `expected_giou_loss` both in https://github.com/facebookresearch/multimodal/blob/afe52dec1987e5a3764539606a55f430994ebbc8/test/modules/losses/test_mdetr_losses.py and test_distribution.py?

File: test/distributions/test_distributions.py [hunk 3]
```diff
@@ -5062,7 +5074,10 @@ def f(*values):
             xfail = [
                 Cauchy,  # aten::cauchy(Double(2,1), float, float, Generator)
                 HalfCauchy,  # aten::cauchy(Double(2, 1), float, float, Generator)
-                VonMises  # Variance is not Euclidean
+                VonMises,  # Variance is not Euclidean
+                Exponential,  # mkl implementation path on intel cpu will produce diffrent results from jit.trace
```
Comment (jgong5) [inline test/distributions/test_distributions.py:5078]: What's the reason MKL implementation produces different result from jit.trace? It looks a regression to me with this PR change.

Comment (CaoE) [reply test/distributions/test_distributions.py]: Jit is supported for ._exponential() and I removed the JIT WORKAROUND for exponential in  [   def rsample(self, sample_shape=torch.Size()):](https://github.com/pytorch/pytorch/blob/67fbd940bae60e4392fb72eb495d51f6e0261260/torch/distributions/exponential.py#L58) .  Now MKL implementation will produce same results.

**Review by jgong5 — APPROVED** [review]

File: test/distributions/test_distributions.py [hunk 4]
```diff
@@ -2411,6 +2411,18 @@ def ref_log_prob(idx, x, log_prob):
...
         self._check_forward_ad(lambda x: x.exponential_())
 
+        def mean_var(lambd, sample):
+            sample.exponential_(lambd)
+            mean = sample.mean()
+            var = sample.var()
+            self.assertEqual((1. / lambd), mean, atol=1e-2, rtol=1e-2)
+            self.assertEqual((1. / lambd) ** 2, var, atol=2e-2, rtol=1e-2)
+
+        for dtype in [torch.float, torch.double]:
```
Comment (jgong5) [inline test/distributions/test_distributions.py:2421]: Since your change also handles types that are not double or float, please cover other types too, e.g., bfloat16 and half.

**Comment by CaoE** [comment]
MKL implementation will also get improvements on AMD machine. I removed the vender check.
Screenshot (369)

**Review by jgong5 — APPROVED** [review]

**Comment by jgong5** [comment]
@malfet We removed vendor checks in this PR for both bernoulli and exponential distribution kernels since MKL should bring perf benefit for x86 arch in general. Could you please review? Also, please advise AMD point of contact to evaluate the perf impact on AMD CPUs. Thanks.

**Comment by CaoE** [comment]
@izaitsevfb 
https://github.com/facebookresearch/multimodal/blob/afe52dec1987e5a3764539606a55f430994ebbc8/test/modules/losses/test_mdetr_losses.py#L115. 
May I know why fixed numbers are used here since `test_exponential_sample` in test_distribution.py already uses `scipy.stats` to test sampling results for exponential. 
The fixed `expected_l1_loss = torch.tensor(0.7721)` and `expected_giou_loss = torch.tensor(1.1768)` is not applicable for new exponential.

**Comment by CaoE** [comment]
@malfet Can we modify the expected loss in https://github.com/facebookresearch/multimodal/blob/afe52dec1987e5a3764539606a55f430994ebbc8/test/modules/losses/test_mdetr_losses.py#L115  to accommodate the new exponential ?

**Comment by izaitsevfb** [comment]
> https://github.com/facebookresearch/multimodal/blob/afe52dec1987e5a3764539606a55f430994ebbc8/test/modules/losses/test_mdetr_losses.py#L115.
> May I know why fixed numbers are used here since `test_exponential_sample` in test_distribution.py already uses `scipy.stats` to test sampling results for exponential.
> The fixed `expected_l1_loss = torch.tensor(0.7721)` and `expected_giou_loss = torch.tensor(1.1768)` is not applicable for new exponential.

Perhaps the test author @ebsmothers can answer that?

**Comment by CaoE** [comment]
@malfet @izaitsevfb As suggested by @ebsmothers,  we change the expected value to match the exact value from the new exponential implementationhttps://github.com/facebookresearch/multimodal/pull/364. Could you please check if this PR will cause internal breakage ?

**Comment by izaitsevfb** [comment]
> @malfet @izaitsevfb As suggested by @ebsmothers, we change the expected value to match the exact value from the new exponential implementationhttps://github.com/[facebookresearch/multimodal/pull/364](https://github.com/facebookresearch/multimodal/pull/364). Could you please check if this PR will cause internal breakage ?

I've manually imported this PR to run the internal tests. For Meta employees, see [D41587318](https://www.internalfb.com/diff/D41587318)

**Comment by malfet** [comment]
@izaitsevfb you need to re-import again, as @CaoE just pushed several commits to the PR after the import.

**Comment by izaitsevfb** [comment]
> @izaitsevfb you need to re-import again, as @CaoE just pushed several commits to the PR after the import.

I think GH timeline is misleading, the commits were made 2 hours ago, before the import. I'll double check via the code.

**Comment by izaitsevfb** [comment]
@CaoE, there are some failures, related to the values, expected by the internal models that I can't share specifics about. Please don't merge this yet. I'm trying to find a PoC that can provide more context.

**Comment by izaitsevfb** [comment]
@CaoE, I've confirmed with the test owners, it's all good. The tests didn't expect the RNG change, but they can be forward fixed. 

For the diff train oncall: please see the details in [D41587318](https://www.internalfb.com/diff/D41587318).

**Comment by CaoE** [comment]
> @CaoE, I've confirmed with the test owners, it's all good. The tests didn't expect the RNG change, but they can be forward fixed.

Thank you @izaitsevfb. May I know if these tests will be fixed for the new RNG in the near future?

**Comment by izaitsevfb** [comment]
> > @CaoE, I've confirmed with the test owners, it's all good. The tests didn't expect the RNG change, but they can be forward fixed.
> 
> Thank you @izaitsevfb. May I know if these tests will be fixed for the new RNG in the near future?

I think there are plans to rewrite the tests to not rely on RNG specifics, but I don't have the specifics. Please don't worry about that, as I've confirmed that the internal test failures are expected and can be forward-fixed.

**Comment by CaoE** [comment]
> I think there are plans to rewrite the tests to not rely on RNG specifics, but I don't have the specifics. Please don't worry about that, as I've confirmed that the internal test failures are expected and can be forward-fixed.

Can we merge this PR first and then fix the tests?

**Comment by izaitsevfb** [comment]
> > I think there are plans to rewrite the tests to not rely on RNG specifics, but I don't have the specifics. Please don't worry about that, as I've confirmed that the internal test failures are expected and can be forward-fixed.
> 
> @izaitsevfb That's great ! Can we merge this PR first and then fix the tests?

Yep, that's what I meant, sorry for not being clear.
